src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/CollectionPropertiesShouldBeReadOnly.cs

```diff
@@ -80,8 +90,29 @@ public static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol
                 return;
             }
 
-            // make sure this property is NOT indexer, return type is NOT array but implement ICollection
-            if (property.IsIndexer || Inherits(property.Type, arrayType) || !Inherits(property.Type, iCollectionType))
+            // make sure this property is NOT an indexer
+            if (property.IsIndexer)
+            {
+                return;
+            }
+
+            // make sure return type is NOT array
+            if (Inherits(property.Type, arrayType))
+            {
+                return;
+            }
+
+            // make sure property type implements ICollection or ICollection<T>
+            if (!Inherits(property.Type, iCollectionType) && !Inherits(property.Type, genericICollectionType))
+            {
+                return;
+            }
+            
+            // exclude Immutable collections
+            // see https://github.com/dotnet/roslyn-analyzers/issues/1900 for details
+            if (property.Type.IsSealed &&
+                property.Type.Name.StartsWith("Immutable", StringComparison.Ordinal) &&
+                property.Type.ContainingNamespace?.ToDisplayString() == "System.Collections.Immutable")
```

Comment:
❗️ I'm not especially happy with this implementation.

1. It fails to account for a property of type `IImmutableList<T>`, which seems to fit directly in the intended scenarios
2. It fails to account for the immutable collections in [tunnelvisionlabs/dotnet-trees](https://github.com/tunnelvisionlabs/dotnet-trees), which indirectly fits the intended scenarios

It seems both cases could be handled by excluding properties whose static type implements an immutable collection interface.

Reply:
Ah good points - I was being lazy to enumerate all the immutable interfaces. I will specifically add checks for the interfaces listed [here](https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable?view=netcore-2.2#interfaces) instead of this heuristic.

Reply:
Fixed in https://github.com/dotnet/roslyn-analyzers/pull/1957/commits/84ba7cd41b7688c86e5b9e569990187b39aa5615